server/src/main/java/org/elasticsearch/index/mapper/vectors/MultiDenseVectorFieldMapper.java

```diff
@@ -0,0 +1,443 @@
+            if (o instanceof Integer == false) {
+                throw new MapperParsingException("Property [dims] on field [" + n + "] must be an integer but got [" + o + "]");
+            }
+
+            return XContentMapValues.nodeIntegerValue(o);
+        }, m -> toType(m).fieldType().dims, XContentBuilder::field, Object::toString).setSerializerCheck((id, ic, v) -> v != null)
+            .setMergeValidator((previous, current, c) -> previous == null || Objects.equals(previous, current))
+            .addValidator(dims -> {
+                if (dims == null) {
+                    return;
+                }
+                int maxDims = elementType.getValue() == DenseVectorFieldMapper.ElementType.BIT ? MAX_DIMS_COUNT_BIT : MAX_DIMS_COUNT;
+                int minDims = elementType.getValue() == DenseVectorFieldMapper.ElementType.BIT ? Byte.SIZE : 1;
+                if (dims < minDims || dims > maxDims) {
+                    throw new MapperParsingException(
+                        "The number of dimensions should be in the range [" + minDims + ", " + maxDims + "] but was [" + dims + "]"
+                    );
+                }
+                if (elementType.getValue() == DenseVectorFieldMapper.ElementType.BIT) {
+                    if (dims % Byte.SIZE != 0) {
+                        throw new MapperParsingException("The number of dimensions for should be a multiple of 8 but was [" + dims + "]");
+                    }
+                }
+            });
+        private final Parameter<Map<String, String>> meta = Parameter.metaParam();
+
+        private final IndexVersion indexCreatedVersion;
+
+        public Builder(String name, IndexVersion indexCreatedVersion) {
+            super(name);
+            this.indexCreatedVersion = indexCreatedVersion;
+        }
+
+        @Override
+        protected Parameter<?>[] getParameters() {
+            return new Parameter<?>[] { elementType, dims, meta };
+        }
+
+        public MultiDenseVectorFieldMapper.Builder dimensions(int dimensions) {
+            this.dims.setValue(dimensions);
+            return this;
+        }
+
+        public MultiDenseVectorFieldMapper.Builder elementType(DenseVectorFieldMapper.ElementType elementType) {
+            this.elementType.setValue(elementType);
+            return this;
+        }
+
+        @Override
+        public MultiDenseVectorFieldMapper build(MapperBuilderContext context) {
+            // Validate again here because the dimensions or element type could have been set programmatically,
+            // which affects index option validity
+            validate();
+            return new MultiDenseVectorFieldMapper(
+                leafName(),
+                new MultiDenseVectorFieldType(
+                    context.buildFullName(leafName()),
+                    elementType.getValue(),
+                    dims.getValue(),
+                    indexCreatedVersion,
```

Comment:
should we also all index_options for quantization: int8_flat, int4_flat?

Reply:
No, we should not. those require a large amount of work and I am doing bare minimum for now. If we want to add automatic quantization here, we will need to rewrite how all the lower flat codecs store vectors as now it would be possible to store more than 2B individual vectors, which consequently means we no longer can have vector ids stored in integers, but then become long values. 

I am going for something simple first for reranking via scripts only.